// Copyright 2019 Shift Cryptosecurity AG
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
package shiftcrypto.bitbox02;

import "common.proto";

import "backup_commands.proto";
import "bitbox02_system.proto";
import "btc.proto";
import "cardano.proto";
import "eth.proto";
import "keystore.proto";
import "mnemonic.proto";
import "system.proto";
import "shamir.proto";

import "perform_attestation.proto";

message Error {
    int32 code = 1;
    string message = 2;
}

message Success {
}

message Request {
    reserved 1, 23;
    oneof request {
        // removed: RandomNumberRequest random_number = 1;
        SetDeviceNameRequest device_name = 2;
        SetDeviceLanguageRequest device_language = 3;
        DeviceInfoRequest device_info = 4;
        SetPasswordRequest set_password = 5;
        CreateBackupRequest create_backup = 6;
        ShowMnemonicRequest show_mnemonic = 7;
        BTCPubRequest btc_pub = 8;
        BTCSignInitRequest btc_sign_init = 9;
        BTCSignInputRequest btc_sign_input = 10;
        BTCSignOutputRequest btc_sign_output = 11;
        InsertRemoveSDCardRequest insert_remove_sdcard = 12;
        CheckSDCardRequest check_sdcard = 13;
        SetMnemonicPassphraseEnabledRequest set_mnemonic_passphrase_enabled = 14;
        ListBackupsRequest list_backups = 15;
        RestoreBackupRequest restore_backup = 16;
        PerformAttestationRequest perform_attestation = 17;
        RebootRequest reboot = 18;
        CheckBackupRequest check_backup = 19;
        ETHRequest eth = 20;
        ResetRequest reset = 21;
        RestoreFromMnemonicRequest restore_from_mnemonic = 22;
        // removed: BitBoxBaseRequest bitboxbase = 23;
        RootFingerprintRequest fingerprint = 24;
        BTCRequest btc = 25;
        ElectrumEncryptionKeyRequest electrum_encryption_key = 26;
        CardanoRequest cardano = 27;
        BIP85Request bip85 = 28;
        ShowShamirRequest show_shamir = 29;
        RestoreFromShamirRequest restore_from_shamir = 30;

    }
}

message Response {
    reserved 3;
    oneof response {
        Success success = 1;
        Error error = 2;
        // removed: RandomNumberResponse random_number = 3;
        DeviceInfoResponse device_info = 4;
        PubResponse pub = 5;
        BTCSignNextResponse btc_sign_next = 6;
        ListBackupsResponse list_backups = 7;
        CheckBackupResponse check_backup = 8;
        PerformAttestationResponse perform_attestation = 9;
        CheckSDCardResponse check_sdcard = 10;
        ETHResponse eth = 11;
        RootFingerprintResponse fingerprint = 12;
        BTCResponse btc = 13;
        ElectrumEncryptionKeyResponse electrum_encryption_key = 14;
        CardanoResponse cardano = 15;
        BIP85Response bip85 = 16;
    }
}
